astronomy/date_body.hpp

```diff
@@ -0,0 +1,781 @@
+                             ? digit_range(digits / 10, 0, size - 1) * 10 +
+                                   digits % 10
+                             : digit_range(digits / 10, begin - 1, size)));
+}
+
+// Returns x * 10 ** count.
+constexpr std::int64_t add_0s(std::int64_t const x, int const count) {
+  return CHECKING(count >= 0, count == 0 ? x : add_0s(x * 10, count - 1));
+}
+
+// Returns |date| advanced by the specified number of |days|. |days| must be
+// nonnegative, and the result must be in the same year.
+constexpr Date add_days_within_year(Date const& date, int const days) {
+  return CHECKING(
+      days >= 0,
+      days == 0
+          ? date
+          : (date.day() + days > month_length(date.year(), date.month())
+                 ? CHECKING(
+                       date.month() <= 11,
+                       add_days_within_year(
+                           Date::Calendar(date.year(), date.month() + 1, 1),
+                           days - month_length(date.year(), date.month()) +
+                               date.day() - 1))
+                 : Date::Calendar(date.year(),
+                                  date.month(),
+                                  date.day() + days)));
+}
+
+// The |day|th day of some |year|.  The resulting date need not be in |year|.
+constexpr Date ArbitraryOrdinal(int const year, int const day) {
+  return day < 1
+             ? ArbitraryOrdinal(year - 1, gregorian_year_length(year - 1) + day)
+             : day > gregorian_year_length(year)
+                   ? ArbitraryOrdinal(year + 1,
+                                      day - gregorian_year_length(year))
+                   : Date::Ordinal(year, day);
+}
+
+// Implementation of class |Date|.
+
+constexpr Date Date::YYYYMMDD(std::int64_t digits) {
+  return CHECKING(digits > 0 && digits < 9999'99'99,
+                  Date::Calendar(digit_range(digits, 4, 4),
+                                 digit_range(digits, 2, 2),
+                                 digit_range(digits, 0, 2)));
+}
+
+constexpr Date Date::YYYYDDD(std::int64_t digits) {
+  return CHECKING(
+      digits > 0 && digits < 9999'999,
+      Date::Ordinal(digit_range(digits, 3, 4), digit_range(digits, 0, 3)));
+}
+
+constexpr Date Date::YYYYwwD(std::int64_t digits) {
+  return CHECKING(digits > 0 && digits < 9999'99'9,
+                  Date::Week(digit_range(digits, 3, 4),
+                             digit_range(digits, 1, 2),
+                             digit_range(digits, 0, 1)));
+}
+
+constexpr Date Date::Calendar(int const year, int const month, int const day) {
+  return CHECKING(year >= 1583 && year <= 9999 && month >= 1 && month <= 12 &&
+                  day >= 1 && day <= month_length(year, month),
+                  Date(year, month, day));
+}
+
+constexpr Date Date::Ordinal(int const year, int const day) {
+  return CHECKING(
+      day >= 1 && day <= gregorian_year_length(year),
+      add_days_within_year(Date::Calendar(year, 1, 1), day - 1));
```

Comment:
It is going to cause civilization to crumble, though perhaps not as much as January 0.5.